source/slang/slang-check-expr.cpp

```diff
@@ -2068,7 +2068,19 @@ IntVal* SemanticsVisitor::tryConstantFoldExpr(
             return nullptr;
         if (!isValidCompileTimeConstantType(substType))
             return nullptr;
-        auto val = tryConstantFoldExpr(typeCastOperand, kind, circularityInfo);
+
+        IntVal* val = nullptr; 
+        if (auto floatLitExpr = typeCastOperand.as<FloatingPointLiteralExpr>())
+        {
+            // When explicitly casting from float type to integer type, let's fold it as an integer value.
+            const IntegerLiteralValue value = IntegerLiteralValue(floatLitExpr.getExpr()->value);
+            val = m_astBuilder->getIntVal(m_astBuilder->getFloatType(), value);
```

Comment:
the type here should be substType, not float type.

Reply:
After replacing it with `substType`, some of tests started failing.

The previous change of mine was copied from how Boolean type was handled in `SemanticsVisitor::tryConstantFoldExpr` ([link](https://github.com/shader-slang/slang/blob/368c2d6eb4c5049dccc1291d3350b4539cdd3c85/source/slang/slang-check-expr.cpp#L1989))
```
    if (auto boolLitExpr = expr.as<BoolLiteralExpr>())
    {
        // If it's a boolean, we allow promotion to int.
        const IntegerLiteralValue value = IntegerLiteralValue(boolLitExpr.getExpr()->value);
        return m_astBuilder->getIntVal(m_astBuilder->getBoolType(), value);
    }
```

So I think it should be `m_astBuilder->getFloatType()` not substType.

Reply:
I think you are right. It should be `substType`.

But then, I am not sure what the existing test started failing.